### DISCUSSION TITLE
FR: recorded files API

### DISCUSSION THREAD

**Discussion by kounoike** [mirakc/mirakc#2055] [Ideas]
（性懲りもなくまた）mirakcのフロントエンドになる録画管理アプリを作ろうとしているのですが、
録画されたTSファイルにアクセスするときにmirakcの設定やdockerのボリュームマウントなど色々と複雑になって頭を悩ませていました。

https://github.com/mirakc/mirakc/blob/main/docs/web-api.md#web-api-endpoints-for-recording ではListing/Getting metadata/Streamingの3つのAPIはout of scopeとなっていますが、この3つのAPI＋削除（schedule.jsonからの削除＆ファイルシステムからの削除）の機能を実装することは出来ないでしょうか？

以前タイムシフト機能を使って全録システムを構築したときは番組終了のイベントを拾って https://github.com/mirakc/mirakc/blob/main/docs/web-api.md#get-apitimeshiftrecorderrecordsrecordstream でTSファイルを取得して処理することで、mirakcとクライアント側の独立性を保ててシステムを構築しやすかったです。
同じ開発体験を通常の録画時にもしたいため、この機能の実装を検討していただけると助かります。

**Comment by masnagam** [comment]
少し考えてみたいと思います．

以下の記述は決してmirakcにapiを追加することを拒否しているわけではありません．
どのようにするのがよいのか考えるための話し合いだと捉えてください．

現状，私自身は

- mirakc
- recman
- redis (トランスコードのためのキュー)

のように複数のサーバーを動かしています．recmanは[recording/simple-rules.js](https://github.com/mirakc/contrib/blob/main/recording/simple-rules.js)を改造したものとなっており，

```js
client.on('recording.stopped', async (schedule) => {
  console.info(`Recording stopped: ${schedule.program.id}: ${schedule.program.name}: ${schedule.state}`);
  // update metadata
  await Deno.writeTextFile(
    `/recording/${schedule.options.contentPath}.json`,
    JSON.stringify(schedule));
  if (schedule.state === 'finished') {
    const job = await addTranscodeJob(client, transcodeQueue, schedule);
    console.info(`Recording stopped: ${schedule.program.id}: ${schedule.program.name}: transcode#${job.id}`);
  }
});
```

のような感じで録画終了後の処理をしています．mirakcを直接改造しなくても，別サーバーに必要なエンドポイントを追加すれば，恐らく欲しい機能は実装できるのではないかと思います．

ただ，この場合はサーバーを２つ動かすことになりますし，mirakcのマウント機能を使ってフロントエンドを動かす場合，same originの制限をいくらか緩和しないといけないだろうとは思います．そう考えると，mirakcにapiを追加することには十分利点があると思われます．

mirakcにapiを追加した場合，恐らく以下のような要望が自然と発生すると想像します．

- 録画データの検索機能
  - 録画数が少ない場合はフロントエンドで処理できますが，多くなってくるとjsで処理するのは難しくなるだろうと思われます（数千件とかでも今のpcなら処理できるかも？）
- 空き容量が少なくなったら古い録画を自動削除
  - 録画中に容量が足りなくなる可能性があるため，これをフロントエンドで実装するのは多分簡単ではないです
- その他，一般的なレコーダーに備わっている機能を追加してほしいという要望

これらはごく自然な要望ですし（私は必要としていませんが），実装するべきなのでしょうが，すべてをmirakcに追加していった場合，メンテできるか少し不安があります．追加するにしても，オーナー制などを導入して機能を実装した人たちだけでメンテしていけるような仕組みを作ってから追加しないと，後で面倒なことになるのではないかと少し心配です．

> schedule.jsonからの削除

録画済みの情報は一時的なもので，録画後１５時間が経過すると自動で削除されます．これはある種のログみたいなもの（録画完了通知みたいなもの）なので，消さなくても大丈夫ではないかと思います．

**Reply by kounoike** [reply]
1つ誤解があるようなのでまず訂正しておきます
「フロントエンド」と書きましたが、今のところブラウザで直接mirakcを呼び出す予定ではなく、一旦サーバ（今度はRust実装をしようかと）で受けるつもりです
最終的にはユーザはブラウザで操作しますが、直接mirakcとは通信しない予定です

**Reply by masnagam** [reply]
だとすると，mirakcにapiを追加しなくても，そのサーバーで一覧や削除機能を提供できそうな気がしますが，それでは良くないのでしょうか？

timeshiftは扱っているファイルの内容が少し特殊なのでapiは必要だと思いますが，録画ファイルは単なるtsファイルなので，ファイルのパスさえ取得できれば，mirakcなしですべて処理できそうな気がします．実際先程記載したrecmanでもそのような処理を行っています．

「同じ開発体験」の「番組終了のイベントを拾って」の部分は多分通常の録画の場合でも同じだと思います．

「TSファイルを取得」はコンテナーに録画用のフォルダーをマウントしている場合，`config.yml`の`recording.basedir`と[POST /api/recording/schedules](https://github.com/mirakc/mirakc/blob/main/docs/web-api.md#postapirecordingschedules)で指定する`options.contentPath`だけでは確かにファイルパスを取得できませんが，パスの`recording.basedir`に相当する部分を特定のパスに置き換えるだけで，別サーバーからアクセス可能なファイルパスに変換できるのではないかと思います．パスの変換自体は難しくないと思いますが，これではだめなのでしょうか？

**Reply by kounoike** [reply]
その場合、mirakcをどういう風に設定（docker composeでどうボリュームマウントしたか、recording.basedirがどうなってるか、録画予約時にどういうファイル名に設定したか（これは今回はそのサーバ側で把握できますが））を把握していないとならず、システム間の結合度が非常に高い状態になっている、と考えています
HTTP APIで録画ファイルの取得・削除が出来るとmirakcがファイルシステムをどう扱っているかに関与せずにシステムが作れる上、ファイルシステムを共有していない状態（例えば録画は自宅だがエンコードをクラウドでするとか）でも良くなりシステム構成の柔軟性が上がります

**Reply by masnagam** [reply]
なるほど，理解しました．

提案されたapiの追加自体は多分それほど難しくないのではないかと思います．
録画済み一覧のための情報を新たに保存しないといけませんが，他と同様にjsonファイルで保存するのが簡単でしょう．

削除時に通知する新しいイベントなどを追加したほうがよいかもしれませんが，これもそれほど難しくはないでしょう．

少し考えてみます．

**Reply by kounoike** [reply]
削除には
- 管理から外す（JSONから消す）だけ
- 実際にファイルも消す
の2通りがあると良さそうですね

録画済み一覧の保存については今まで通りのschedule.jsonだけにして、設定ファイルに「何時間後に消す」or「消さないようにする」の設定を付けるのでも良いかもしれません

後は、JSONが巨大になるとたまたま保存のタイミングで強制終了とかあると不安ですので、SQLiteあたりの導入も考えた方が良いかもしれませんね。まあ、とりあえずはJSONで良いと思います

**Reply by masnagam** [reply]
ああそうか，特定のフォルダーにファイルが入っているんだから，それを`ls`すればいいだけか．
メタ情報はtsから再度取り出すこともできるけど，念の為，別ファイルに書き出せば，一覧を管理するファイルやデータベースを別途用意しなくても良さそうです．

もっと簡単に実装できそうな気がしてきました．

**Reply by masnagam** [reply]
強制終了がsigtermであれば，録画途中のものはどうにもなりませんが，完了後のメタデータ保存とかなら，それが完了するまで待てると思います．
sigkillとかが来たら，それは送ったほうが悪いので救済不能ですが．

**Comment by kounoike** [comment]
> 録画データの検索機能

確かにあるといいなとは思いましたが、検索の実現には色々あるのでmirakc自体が提供する必要はないかなと思いました
単純な部分一致、正規表現、あいまい検索、embeddingによるスマート検索、EPG/ARIB字幕/文字起こしからの内容での検索、などなど色々考えられますが、それはmirakcの外側で提供することかなと思いますし、そういう管理アプリを作ろうと思っています

> 空き容量が少なくなったら古い録画を自動削除

これも分かるのですが、最低限にして空き容量の取得だけが出来るようになれば、クライアント側で監視して必要に応じて削除処理をしていけば良いのかなと思います（そのためにも削除APIが欲しい）

> 15時間が経過すると自動で削除

あれ、そうなんでしたっけ・・・
そうなると一覧の元になるデータがなくて困ったことになりますね・・・
もしかすると提案したAPI実装するとしたら、ここが一番大事な変更点になるかもしれないです

**Reply by masnagam** [reply]
> それはmirakcの外側で提供することかなと思いますし、そういう管理アプリを作ろうと思っています

私も同じ意見です．ただ，録画件数が多くなるとブラウザーのjsで処理するのはメモリー使用量的に困難となるかもしれません．その場合，フロントエンドから（mirakcとは別の）サーバーにリクエストを送って，サーバー側で実際の検索処理などを行うことになります．一般的には多分サーバーでデータベースなどを使って録画を管理するのだろうなとは思っています．mirakcでは面倒なので全てjsonファイルに情報を書き込んでいますが．

> そうなると一覧の元になるデータがなくて困ったことになりますね・・・

`recording.stopped`イベントがあるので，これを受信したときに録画情報をmirakcから取得すれば大丈夫です．イベント受信ができなかったとしても，スケジュールに記録されている情報はすぐに消えることはないので，最悪定期的に取得すれば情報は回収できると思います．

**Reply by kounoike** [reply]
私がやろうとしていることは今のところ録画が終わったら即座にmirakcからファイルを回収して別の場所で管理するようにしようかと考えていますが、mirakc自体の機能として録画済みファイルの一覧機能を提供する場合には以前に録画したファイルをすべて覚えていないとならず大変かなと思いました

**Reply by kounoike** [reply]
（本筋から脱線してますが）

> サーバー側で実際の検索処理などを行うことになります

最近文字起こしやドキュメント検索などの技術の進歩が活発なので、その成果を取り入れて大量の録画の中から欲しい情報を見つけやすく出来ると面白いかな、と考えています
例えば[meilisearch](https://www.meilisearch.com/)にEPGから得た情報、ARIB字幕や文字起こし結果などを入れておくだけでも面白そうかな、と

**Reply by masnagam** [reply]
今なら自然言語で検索とかするのが良いのでしょうね．
and/or条件とかやり始めると面倒ですし，google検索とかで機能は提供されてますが，自分なんかほぼキーワードの羅列でしか検索してません．
自分が特別だとは思わないので，多分多くの人がand/orなどの機能は使ってないでしょうね．

> 以前に録画したファイルをすべて覚えていないとならず大変かなと思いました

数が多くなるとjsonでは難しくなってきますね．データベースを使いたいところですが，導入するべきか少し悩みます．
メタデータを別ファイルにすれば，録画一覧を保存するjsonのサイズは小さくなるので，とりあえずこの方法で何とかなるかというところですね．
無理そうならデータベースを素直に導入することになると思います．

**Reply by kounoike** [reply]
and/or条件の検索式の標準化をして欲しいです・・・
システムごとに違ってて検索の仕方が分からないので

個人的にはmirakcは単体で動いていて欲しいのでDB導入するならsqlite推しです・・・

**Reply by kounoike** [reply]
多分真面目にテーブル定義しないで、
- id
- json_str
の2カラムだけのテーブルで十分そうな気がします

**Comment by masnagam** [comment]
スレッドが別れてしまったので，新たに作成します．

> 削除には

ファイルを１つ削除するのも２つ削除するのも同じことなので，対応は難しくないと思います．

録画用フォルダーにはmirakcが作成するファイル以外を入れないことを前提条件として，録画一覧jsonやデータベースを導入しない実装をまずは考えてみようと思います．

**Reply by masnagam** [reply]
apiを追加するので，多分マイナーバージョンを更新します．
apiを追加したとしても，しばらくはreleaseには反映されないので，フロントエンドの開発にはmainからビルドしたmirakcを使ってください．

**Reply by kounoike** [reply]
> ああそうか，特定のフォルダーにファイルが入っているんだから，それをlsすればいいだけか．

contentPathでディレクトリ掘ってることもありますよ・・・？

クソデカschedule.jsonになることを避けるなら、schedules/XXXXX.json, YYYYYY.jsonって感じでprogram_id辺りで採番したファイル名にため込むとかもアリなように思いました

**Reply by kounoike** [reply]
（番組タイトルからファイル名付けてたら勝手にディレクトリ掘られてました🤣）

**Reply by masnagam** [reply]
そうですね，パスは自由に指定できます．
しかし，それはあくまでPOSTに指定した`options.contentPath`が原因なので，指定しない限りはそうはなりません．

`options.contentPath`に制限を追加するか，一覧するファイルの拡張子を`config.yml`などで指定させれば，それを元にファイルを集めることが可能だと思います．

**Reply by masnagam** [reply]
#2056 を追加しました．

**Reply by kounoike** [reply]
フォルダ掘る機能を、例えば `YYYY/YYYY-mm-dd/番組タイトル.ts` にしてる人とかもいると思うので、制限追加はちょっと・・・
あとファイルパスをリクエストに含めるのもなんか気持ち悪い（言語化出来てなくてすみません）ですね

👆の特定ディレクトリに `{id}.json` 作っていくのはどうですか？
例えば `GET recording/records/{id}/stream` のときもJSON見に行く→contentPathが分かる→ファイルが読めるし、`GET recording/records` はそのディレクトリ以下のJSON全部読んで繋いで返すだけなので・・・

**Comment by kounoike** [comment]
（ここまでやってきて言うのもなんですが）
スレッドリプライはあまり使わないようにして、縦に伸ばしていった方が逆に読みやすいかもしれないですね

**Reply by masnagam** [reply]
同意します

**Reply by kounoike** [reply]
それをリプライで言いますか？（笑）

**Comment by masnagam** [comment]
> 例えば YYYY/YYYY-mm-dd/番組タイトル.ts にしてる人とかもいると思うので

この場合でも，`*.ts`を収集対象とすれば行けるのではないかと思います．

> あとファイルパスをリクエストに含めるのもなんか気持ち悪い

なるほど，idを使う場合，idとパスのマッピングが必要になり，結局管理ファイルが必要になるのか．
当初これらのapiを追加しないつもりだったので，`options.contentPath`を追加して好きな名前を作れるようにしたのですが，面倒なことになりましたね．

> 👆の特定ディレクトリに {id}.json 作っていくのはどうですか？

これはいいアイディアです．`{id}.json`にメタデータも入れておけば，ファイルの数も２つのままで増えませんし．

**Comment by masnagam** [comment]
メタデータには録画予約情報も入っているから，当然パスも入ってますね．

**Comment by kounoike** [comment]
> メタデータには録画予約情報も入っているから，当然パスも入ってますね．

はい。ほぼschedule.jsonに入ってるうちの1レコードがファイルに切り出されたもの、というイメージでした
GETでそれがそのまま返ってくれば十分かなと思います

**Comment by masnagam** [comment]
`{id}.json`は別途管理用のフォルダーを作って，そっちに保存します．

録画ファイルを手動で消されたりすると内部矛盾が発生しますが，まあ何とかなるでしょう．

**Comment by masnagam** [comment]
実際の作業は作業ブランチで行います．

PRを作成するので問題ないかコードレビューをお願いしたいです．
rustのコードはよくわからないという場合には，apiのみの確認だけでも十分です．

**Comment by kounoike** [comment]
> `{id}.json` は別途管理用のフォルダーを作って，そっちに保存します．

そうですね。jsonがあって.tsが無いときには404か500とかで返る感じでしょうか
そういう「tsがあってjsonが無いときにそのjsonを消す」機能もあればベターでしょうけど、無くてもまあ・・・？
`DELETE recording/records/{id}` のときにエラーにならずにjsonだけを消せればとりあえずは良いと思います！

**Comment by masnagam** [comment]
> DELETE recording/records/{id} のときにエラーにならずにjsonだけを消せればとりあえずは良いと思います！

確かに，それだけでいけるかもしれませんね．

**Comment by masnagam** [comment]
今日はもうこれ以上返信しませんが，なにかあればいつでもここにコメントを書いてください．

**Comment by kounoike** [comment]
> > DELETE recording/records/{id} のときにエラーにならずにjsonだけを消せればとりあえずは良いと思います！
> 
> 確かに，それだけでいけるかもしれませんね．

GET系の何かで「あ、jsonがあってtsが無い状態だな」って分かるようなレスポンスになってるとやりやすいですね
`HEAD recording/records/{id}/stream` でtsの有無が確認出来るとか

> 今日はもうこれ以上返信しませんが，なにかあればいつでもここにコメントを書いてください．

リアルタイムでレスが返ってきたので助かりました！ありがとうございました！

**Reply by kounoike** [reply]
👆があると一覧取得してjson/tsの対応取れてるか全部チェックして対応してないのは消していく
みたいなツールをクライアント側で書けますね、という話です

**Comment by masnagam** [comment]
> GET系の何かで「あ、jsonがあってtsが無い状態だな」って分かるようなレスポンスになってるとやりやすいですね

`GET recording/records/{id}`で返すモデルに`size`を追加します．ファイルがないと`null`になります．

> HEAD recording/records/{id}/stream でtsの有無が確認出来るとか

`stream`系のエンドポイントはすべて`GET/HEAD`を実装しているので，`HEAD recording/records/{id}/stream`も追加します．

**Comment by masnagam** [comment]
とりあえずapiだけ追加しました．
https://github.com/mirakc/mirakc/tree/feat-recording-endpoints

実装はダミーなのでそのまま使うことはできません．
多分stub（mock）を書いてテストをすると思うので，その参考にしてください．

足りない機能があれば，ここでコメントしてください．

**Reply by masnagam** [reply]
ブランチをcheckoutしてビルドして動かせば，`localhost:40772/api/debug`からもドキュメントが読めます．
ただ，上で書いたようにダミー実装なので，swagger uiからのapi呼び出しは機能しません．

**Comment by kounoike** [comment]
ありがとうございます！
まだ動かしてないですが、エンドポイントとしては良さそうに思います

録画終了前にstreamにアクセスしたときの挙動が気になりました
欲を言えば録画終了前にstream読むと、録画済みの部分は一気に来た上で残りは受信するごとに徐々に送られてくる、みたいなことが出来ると放送中に直でオブジェクトストレージに突っ込んだり、リアルタイムでエンコードして行ったりということも出来るかな、と思いました
・・・が、まあとりあえずはシンプルに録画終了後からしか読めない、でも良いと思います

**Comment by masnagam** [comment]
> 欲を言えば録画終了前にstream読むと、録画済みの部分は一気に来た上で残りは受信するごとに徐々に送られてくる

最終的には録画開始時にrecordを作成するので，録画中のデータを読み出すことも可能になります．
ただ，録画データファイルのEOFに達するとfile readerは閉じてしまうので，その時点でhttp chunk responseは終了してしまうのではないかと推測します．

現状，単純にファイルをopenしてreadしていますが，`tail -c +0 -f`でファイルを読み出せば所望の動作にできるとは思います．この場合，自然とEOFにはならないので，録画終了時にhttp chunk responseを終了させる必要があります．

**Comment by masnagam** [comment]
どうもrange requestsが謎実装になっていることがわかったので，しばらくの間は使わないでください．チケットを作ってあるので後で修正します．

**Comment by masnagam** [comment]
データベースなしでrecord idがユニークであることを保証するのは面倒なので，record idの型を数値から文字列に変更しようと思います．
同じ番組を同時に複数録画することはできない仕様になっているので，program.idとtimestampからrecord id文字列を作ることにします．
多分固定長の文字列になると思います．

**Comment by masnagam** [comment]
タイミングなどの問題から，recordの作成・更新・削除のイベントを別途追加します．
`recording.started`の時点ではrecordに安全にはアクセスできません．
recordへのアクセスはrecord作成通知後に行う必要があります．

**Comment by masnagam** [comment]
@kounoike 実験に使えるレベルまで実装が完了しました．ドキュメントやテストを追加することはあると思いますが，不具合がない限りapiが変わることはないと思います．

差分確認用に PR #2073 を作成してありますが，まだ作業は完了していません．完了したらapiなどの確認のためレビュー依頼を出します．

**Reply by masnagam** [reply]
PRでのチェックが失敗していますが，ローカルではテストはすべてパスしています．
多分，環境の違いが原因だと思います．

**Comment by masnagam** [comment]
@kounoike githubでreviewerに設定するためにはメンバーとしてどこかに追加しないといけないみたいなので，とりあえずreadでmirakc/mirakcに追加してみました．この機能を使ったことがないのでわからないのですが，もしかするとteam設定とかをいじらないと無理かもしれません．あまりにも面倒な場合にはgithubの機能を使うのを諦めることにします．

**Reply by kounoike** [reply]
invite受けておきました
レビューコメントも出来てるようです
（意図せずパーミッションが空きすぎていて、何か変なところに書けたりしないかとかは未確認です）

**Comment by kounoike** [comment]
まだ本格的に使い始めていないですが、ちょっと試したところ、「contentPathに興味がない（実体が何て名前のファイルでも良い）のに、必須項目で面倒くさい」と感じました。
contentPathを省略した場合はmirakc側が勝手に名前つけてくれると嬉しいな、と思います。
まあ、機械的に<programId>.tsとかにすれば良いのでそこまで優先度高い話ではないですが、一応お伝えしておきます。

**Reply by masnagam** [reply]
> 機械的に.tsとかにすれば良いのでそこまで優先度高い話ではないですが、一応お伝えしておきます。

mirakcはもともと録画ファイルを隠さず，ユーザーが直接ファイルにアクセスすることを想定したapiだったという理由からconentPathが必須となっていました．
`/api/records`利用時は直接録画ファイルに触れる機会が減るので，optionalに変更しても良さそうですね．
recordファイル名と同じにすれば良いと思うので，optionalに変更しましょう．

**Reply by masnagam** [reply]
`/api/records`無効化時はこれまで通りcontentPathは必須とします．

**Reply by masnagam** [reply]
90aba211f89876473ca2101486cdcb23e4391acf で実装しました．

`Record.options.contantPath`ではなく，`Record.contentPath`を使ってください．
`Record.options.contantPath`は録画時に指定した値なので，`contentPath`を指定しなかった場合，`undefined`となります．

**Reply by kounoike** [reply]
ええと、ちょっと分からなかったのですが、

- POSTするときにoptions.contentPathを指定しなかった場合
　　- Record.options.contentPathは空
　　- Record.contentPathはmirakcが自動で決めたファイルパスが入っている
- POSTするときにoptions.contentPathを指定した場合
　　- Record.options.contentPath、Record.contentPathともに指定した値が入っている

という認識で大丈夫でしょうか？

**Reply by masnagam** [reply]
概ねあっていますが，以下に正確に書きます．

- POSTするときに`options.contentPath`が`undefined`（つまりcontentPathのプロパティ自体設定しなかった）の場合
  - `Record.options.contentPath`は`undefined`
  - `Record.contentPath`はmirakcが自動で決めた相対ファイルパスが入っている
- POSTするときに`options.contentPath`が`null`の場合（これを指定することは稀だと思いますが）
  - `Record.options.contentPath`は`null`
  - `Record.contentPath`はmirakcが自動で決めた相対ファイルパスが入っている
- POSTするときに`options.contentPath`を指定した場合
  - `Record.options.contentPath`は指定した値
  - `Record.contentPath`は`..`などを解決した相対ファイルパス

`Record.options.contentPath`を書き換えてしまうと，ドキュメントの説明が少し面倒なことになるので，`Record.contentPath`を追加しました．
`Record.options`は`RecordingOptions`型なのですが，その説明が埋め込み先依存で変わってしまうためです．

`config.recording.records-dir`が指定されていない場合は，`options.contentPath`はこれまで通り必須項目になります．

**Comment by masnagam** [comment]
8f97c45775a576fa365b5033d7a8c516dade9826 にて`recording.record-saved`イベントに`recordingStatus`を追加しました．
`recordingStatus`を確認するだけの目的で`GET /api/records/{id}`を呼び出す必要はなくなります．

**Comment by masnagam** [comment]
今更で申し訳ないのですが，recordの構造を以下のように変更しようと思います．

```jsonc
{
  "id": ".."
  "program": { .. },
  "service": { .. },
  "tags": [ .. ]
  "recording": {
    "options": /* record.options から移動 */,
    "status": /* record.recordingStatus から移動 */,
    "startTime": /* record.recordingStartTime から移動 */,
    "endTime": /* record.recordingEndTime から移動 */,
    "duration": /* record.recordingDuration から移動 */
  },
  "content": {
    "path": /* record.contentPath から移動 */,
    "type": /* record.contentType から移動 */,
    "length": /* record.contentLength から移動 */
  }
}
```

**Reply by masnagam** [reply]
675016e5aa5900c25fa29e6a349f6629b73bf838 で移動しました

**Comment by masnagam** [comment]
実装は終わりました．
不具合や要望がなければこれ以上修正することはありません．
確認が終わったらPRを承認してください．

**Comment by kounoike** [comment]
`/events` で、`epg.programs-updated` は接続時に全チャンネル分流れてきますが、同じように `recording.record-saved` も全録画の分流すというのはどうでしょうか？
前者は多分起動の順番などで受け取れなかったときの救済処置だと思うのですが、後者でも同様に受け取れなかったときのリカバーがあったら良いかなと思っての提案です。
クライアント側の実装面でも録画が終わるときしか流れてこないと動作確認が辛いというのもあります。

**Reply by kounoike** [reply]
まあ、record全部流すのはだいぶ重いかもしれないので悩ましいところですけど・・・

**Reply by masnagam** [reply]
`egp.programs-updated`と同様にするなら`async`を使うのでそれほど負荷は大きくない気がしますが，数が問題ですね．．
https://github.com/mirakc/mirakc/blob/ec039c4cd22c83b0ee21099fc35872ef598e3c22/mirakc-core/src/epg/mod.rs#L760

`egp.programs-updated`はチャンネル数程度なので上限はないものの，現状（また将来も）多くとも100以下でしょうが，`recording.record-saved`は数千となる可能性があります．

実際の動作を確認したところ，１つのイベント送信は1ms以下でした（MeLE Quieter 3C 有線接続）．
つまり，1000程度であれば1s以下ですべてのイベントが送受信されます．
ホップ数が多くなるともう少し時間がかかるかもしれませんが，イントラネット以外での運用は禁止なので，ホップ数の多いケースは無視してよいだろうと考えます．

もともとイベントは大量に送っても良いように，idのみ送るなどデータ量を絞ってあります．
多分`/events`接続時に`recording.record-saved`を全録画の分送っても大丈夫でしょう．

**Reply by kounoike** [reply]
設定でオン・オフ出来ると良いかもしれないですね

**Reply by masnagam** [reply]
f06a2255d69f71cb1c796e9f6b33d5b5dbb4da47 で対応しました．

オン・オフは実際に問題となったときに追加することにします．

**Reply by kounoike** [reply]
そうですね、この機能を使う人が他にどれだけいるのかという状態ですから今はオンだけで良いと思います
確認したらApproveしておきます

**Comment by fetburner** [comment]
@masnagam 
お世話になっております。こちらの discussion で検討されていた録画関連の API についてなのですが、リリース予定について伺いたいです。
現在 https://github.com/fetburner/rokuban で mirakc と協調して動作する録画サーバを開発しているのですが、エッジ側に置いた mirakc の保持している録画データにアクセスするため、次の API を利用できないかと考えています。

- GET /api/recording/records
- GET /api/recording/records/{id}
- GET /api/recording/records/{id}/stream
- DELETE GET /api/recording/records/{id}

上記 API は main では実装済みですが、現時点の最新リリース 3.4.81 の方には含まれていないものと存じています。

もし良ければ、録画関連の API を安定版にリリースする予定、またリリースブロッカーが存在しているようであればそれについて教えていただけないでしょうか。
こちらで開発している録画サーバでサポートする mirakc のバージョンを決めるため、見通しだけでも分かると助かります。

**Reply by masnagam** [reply]
まず， `3.*` にマージすることはありません．
リリースする場合には `4.0.0` になります．

api妥当性の検証で停止しています．
つまり，apiは確定していません．
もし追加・修正の要望があれば受け付けます．

api妥当性の検証が完了したら， `4.0.0` としてリリースすると思いますが，時期の明示は困難です．

バージョンチェックのみの問題であれば， `main` ブランチでは `4.0.0-dev.0` としてあるので，semverでmajor versionが4以上と条件を設定すれば  `mirakc/mirakc:main` でも動作すると思います．

**Reply by fetburner** [reply]
ご回答ありがとうございます。API のリリースが 4.0.0 以降になる点、承知致しました。
こちらの開発している録画サーバではサポート対象を mirakc 4.0.0 以降（当面は `mirakc/mirakc:main`）と明記し、頂いた semver の判定条件を使わせて頂きます。

もし API の追加や修正を受け付けて頂けるのでしたら、要望と言うほど強いものではないのですが、`GET /api/recording/records` にS3 の `GetObject` でサポートされているチェックサムを返す機能（リクエストする際に`x-amz-checksum-mode: ENABLED` を指定すると `x-amz-checksum-crc32` 等のレスポンスヘッダでチェックサムが返ってくる）に相当するものを検討して頂けると幸いです。

https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#API_GetObject_ResponseSyntax

mirakc の扱う録画データはそれなりに大きなものなので、転送時にアプリケーション層でも一応誤りを検出してリトライできる手段を確保できればと考えております。

**Reply by masnagam** [reply]
なるほど．`content.sha256`とかでいいですか？録画終了時にsha256を計算して設定し，それまではundefinedとなります．
値の設定は一度だけで，仮に録画データが書き換えられたとしてもsha256値を再計算・再設定はしません．

この仕様で問題なければ追加します．

jsonのデータ形式が変わるので，テストする場合は修正前のデータを破棄するか，手動でsha256値を設定してもらうことになります．

**Reply by masnagam** [reply]
そうか．`${content.path}.sha256`とか別ファイルにしておけば，jsonデータそのままで拡張できるのか．．
少し考えてどっちにするか決めますが，外部仕様はどちらの方式でも変わらないので，先程の仕様で問題なければ機能を追加します．

**Reply by fetburner** [reply]
はい、ハッシュ値さえ得られれば良いので、`content.sha256` でも `${content.path}.sha256` でも、実現の容易な形式で大丈夫です。
録画終了時に SHA-256 のハッシュ値を設定し、録画データが書き換えられても再計算しない仕様で問題ありません。

**Reply by masnagam** [reply]
#3077 で対応しました．

sha256の結果を別ファイルにするとファイル読み込み回数が増えるので，record.jsonに含めることにしました．そのため，record.jsonの形式が変わります．

**Reply by masnagam** [reply]
`/api/recording/records`を使っているようですが，個人的にはrest apiを使わず，sseのイベント通知と録画ファイルへの直接アクセスを使ったほうが，アプリ実装の自由度が高くなるので好ましいと考えています．

要望がありapiを実装しましたが，api化は機能に対する「ある種の制約」を与えることになります．例えば，sha256ではなくblake3を使いたい場合，当然apiを修正しない限り機能を追加できません．一方で，`api/recording/records`を使わず，録画ファイルにアクセスする場合は，このような制約はなく，sse経由で録画完了イベント受信時に希望するハッシュ関数でハッシュ値を計算できます．

パフォーマンスについての懸念もあります．録画データを読み込む場合，ファイルを読み込むほうがhttpストリーミグより効率がよいです．利便性のためどこかでhttpストリーミングするほうがよいですが，それはクライアントに直接対面しているサーバーでのみ行うのが理想です．しかし，rest apiを使うと直接mirakcからクライアントにhttpストリーミングしない限り，余計なhttpストリーミング処理が入ります．nfsなどを使えば効率的にファイルを共有できるので，パフォーマンスの観点からもあまりrest apiの使用はおすすめしません．

録画ファイルを直接使うことで，メタ情報（番組情報やパケットのドロップ量など）の抽出やトランスコードなどもアプリ側で好きなように実装できます．

**Reply by fetburner** [reply]
ご対応頂きありがとうございます。
`content.sha256` の追加、また REST API を用いる場合の留意点についても、丁寧にご説明頂き感謝致します。
頂いた観点は録画サーバの設計を検討する上で参考にさせて頂きます。

**Reply by masnagam** [reply]
@fetburner 
#3077 をマージしました．
今夜`main`タグのdockerイメージが作成されます．
